### ISSUE TITLE
Return async responses

### ISSUE THREAD

**Issue by ebkalderon** [kpcyrd/chrootable-https#10]
It would be nice if `Client::get()` and other request methods returned a `ResponseFuture` for use with async runtimes, like the current `hyper::Client` does for `tokio`. This would be great for easily managing multiple concurrent downloads and measuring the progress of individual streams. This crate already depends on `tokio`, so I imagine it would be technically trivial to expose the asynchronous interface underneath somehow.

**Comment by kpcyrd** [comment]
Correct, I'm using the async apis internally but decided to abstract them so I don't have to deal with them in my other projects. I think a `ResponseFuture` can be introduced in a non-invasive way regardless by implementing a method on the future that turns it back into synchronous code.

If you're interested working on this I can try to assist you.

Keep in mind that there's currently a hack in the codebase that you would probably have to resolve to make it truely asynchronous:

https://github.com/kpcyrd/chrootable-https/blob/f5e04aa4c649b8ea4c4b21506dcbc88985ca43a2/src/connector.rs#L78-L108

The dns name has already been resolved at this point, there is an asynchronous attempt that deadlocked when I wrote it. The pre-resolved records are filled-in over here:

https://github.com/kpcyrd/chrootable-https/blob/f5e04aa4c649b8ea4c4b21506dcbc88985ca43a2/src/connector.rs#L25-L40

**Comment by ebkalderon** [comment]
Hi there! Thank you for all the helpful information. I have completed the work for this feature in my [personal fork](https://github.com/ebkalderon/chrootable-https/tree/async-responses) by uncommenting your async example, updating some things where they could not compile. I have fully updated the `Connector` and I have verified that the tests pass and example program runs without deadlocks on both Linux and macOS.

With that said, I quickly ran into an issue updating the `Client` to return a `ResponseFuture` which stems from the architecture of the included synchronous `Resolver`. This `Resolver` internally creates and destroys its own `tokio::Runtime` every time `resolve()` or `resolve_with()` is called. This prevented me from adding a `wait_for_response()` method to `ResponseFuture` which uses a `tokio::Runtime` to drive itself to completion, as spawning a `Runtime` inside of another `Runtime` causes a panic.

As such, I decided to update the `get` example to use the included `AsyncResolver` instead, just to test out my async implementation, because at least it does not spawn a `tokio::Runtime` inside of itself. However, it turns out that `AsyncResolver` doesn't implement `DnsResolver`, making it unusable. There are no unit or integration tests making use of the `AsyncResolver` either, and the inner fields differ from the standard `Resolver`. It appears to me that this resolver was never actually usable with `Client`, so I completely removed it. In its place, I refactored `Resolver` to be asynchronous and changed `DnsResolver` return a `Resolving` 
future instead of a `Result`.

With the changes in the related PR (#11), the library is now entirely asynchronous, both under the hood and in its public interface, matching the user experience of `hyper`. Unlike `hyper` though, each `Client::get()` request can be made synchronous by chaining a simple `.wait_for_response()` at the end.
